### PR TITLE
[WIP] Proposed revision to sample_posterior_predictive()

### PR DESCRIPTION
**Pull request by rpgoldman**
Most other sampling functions take varnames, rather than var objects. Extend the set of parameters to accept varnames as an alternative to vars, preserving backwards compatibility.
Also revise the docstring, to clarify the return type and add type comments.

### CODE REVIEW

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py]: No need to comment this out. Type hints are ok because they're supported by python 3.5. What we don't support is variable annotation.
However, the line is too long. Try splitting it with the `black` package. Don't `black` the entire file, just the `sample_posterior_predictive` function.

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py:1053]: You should add a note saying that it's deprecated.

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py:1094]: Add an `else` condition that says that `vars` is deprecated in favor of `varnames`

**Review by lucianopaz — CHANGES_REQUESTED** [review]
Overall, I like that this unifies the API consistently. IMHO, I would have left `vars` and deprecated `varnames`. However, I would like to hear what others think about this. Maybe `varnames` is safer because you ensure that the tensors and RVs are part of the model, whereas, `vars`could hold anything out of the model's context.

Finally, if this goes through, you should add a line at the end of the deprecations section of release-notes.

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py]: Also remove unused imports to fix lint

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: Actually, this type hint functions despite being commented out. I'm not sure what the alternative to this form of type annotation for pre-3.6 Python.
So can I use this form in 3.5?

``` python
def sample_posterior_predictive(trace, 
                                                    samples: Optional[int]=None, 
                                                    model: Optional[Model]=None,
                                                    vars: Optional[Iterable[Tensor]]=None,
                                                    varnames: Optional[Iterable[str]]=None,
                                                    size: Optional[int]=None,
                                                    random_seed=None,
                                                    progressbar: bool=True) -> Dict[str, np.ndarray]:
```
I'm afraid I don't understand what is and isn't permitted in 3.5, since I went straight from 2.7 to 3.6.
[Nor, obviously, do I understand what the heck Markdown will do with my whitespace!]

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py:1053]: Oh, are you suggesting I should deprecate this? I wasn't going to go that far -- just add `varnames` as an easier-to-remember alternative. OK, will do.

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py:1094]: Will do.

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: I'm not sure what to do about this -- these imports **are** used by mypy, AFAIK.  I don't know how the checker finds type definitions without imports.
I'm happy to do what's right here, but once again, I don't understand how 3.5 functions.

File: pymc3/sampling.py
Comment (lucianopaz) [reply pymc3/sampling.py:1053]: I think that supporting two parametrizations is just redundant and I think that we should opt for one or the other. @twiecki, @junpenglao, what do you think?

Comment (rpgoldman) [reply pymc3/sampling.py:1053]: I figured you would prefer I not break old code, including example notebooks, that use the old `vars` parameter.
I think `varnames` is better because it preserves the abstraction around the model object, but I could be wrong.

File: pymc3/sampling.py
Comment (lucianopaz) [reply pymc3/sampling.py]: Type hints are specified in [pep484](https://www.python.org/dev/peps/pep-0484/). How to annotate callables is [here](https://www.python.org/dev/peps/pep-0484/#callable). What you suggest seems to be variable annotation instead

Comment (rpgoldman) [reply pymc3/sampling.py]: PEP484 contains this example, which is of the same form as what I gave in my response:

```
def dilate(v: Iterable[Tuple[T, T]], scale: T) -> Iterable[Tuple[T, T]]:
```

I have *also* type hinted local variables, but I moved all that to comments. So I think the example above should be ok with 3.5, if it supports PEP484.

File: pymc3/sampling.py
Comment (lucianopaz) [reply pymc3/sampling.py]: Travis' logs say

> ++++pylint --rcfile=.pylintrc pymc3
> ************* Module pymc3.sampling
> pymc3/sampling.py:1:
[W0611(unused-import), ] Unused Optional imported from typing

File: pymc3/sampling.py
Comment (lucianopaz) [reply pymc3/sampling.py]: Ah, great!

**Comment by lucianopaz** [comment]
I think that this change is nice but I think we should discuss two things:

1. Whether we want to change the function parameters to make the API more consistent.
2. Whether to add a very small test environment for python 3.5, so we ensure that we build and at least sample correctly in that version of python. After #3465 and the variables annotations problems, I think that we need this.

@twiecki, @junpenglao, @ericmjl, @ColCarroll

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: @lucianopaz At least for now I have 'fixed' this by telling pylint to ignore those imports.  I haven't been able to find an alternative solution, because I don't know otherwise how mypy is to identify the types.  I think what we have now is an acceptable expedient, unless someone can let me know of a better solution. I'll look for some advice myself.

**Comment by rpgoldman** [comment]
The deprecation warning was too aggressive.  Fixing that now.

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py]: Use the [warnings package](https://docs.python.org/3.6/library/warnings.html) instead of raising the `DeprecationWarning`. Also, this warning should be raised if `varnames is None` and `vars is not None`.

Comment (rpgoldman) [reply pymc3/sampling.py]: Got this now, I believe.  Waiting for the tests to complete.
There's enough uninteresting history on this topic branch that it's probably worth squash-merging it, or I could clean up the history.  Need to rebase it, anyway.

**Comment by fonnesbeck** [comment]
Its not clear (to me) that these need to be unified. Sampling takes place in a model context, so it makes sense to pass the PyMC variables directly. Whereas, plotting occurs outside the model context, so the variables are not generally available. To me, there is a clear line between the two, because they are associated with distinct tasks in different phases of the modeling workflow.

**Comment by rpgoldman** [comment]
> Its not clear (to me) that these need to be unified. Sampling takes place in a model context, so it makes sense to pass the PyMC variables directly. Whereas, plotting occurs outside the model context, so the variables are not generally available. To me, there is a clear line between the two, because they are associated with distinct tasks in different phases of the modeling workflow.

That may be true, but in addition to being confusing, and having the potential for error that @lucianopaz points out, there's also the issue that since the user is typically working with variable names, they just end up having to write code that looks like:

```
vars=(model[x] for x in name_list)
```
...at least that has been my experience using this API (more accurately, I do that *after* I have encountered the wrong argument error! 😜).  This is inconvenient and potentially error-prone.

Typically I write a function that builds the model and I don't "hang onto" the variable objects, and it's easier to work with the variable names.

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py]: Better to just not import Any. It's only used in the commented variable annotation

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py]: If you remove any from this import, does pylint complain? We prefer to not ignore lint

File: pymc3/sampling.py
Comment (lucianopaz) [reply pymc3/sampling.py:1094]: This can become an else clause

**Comment by lucianopaz** [comment]
@fonnesbeck, sampling from the posterior predictive should also be done within the model's context. Once we get the sample, we can plot outside the context. I'm also hesitant of making this change so I labeled it for discussion.

I now think that using names is safer because that way, we ensure that the variables are defined in the model. If we accept tensors, the users could mix-up tensors that come from two separate models, and could make a strange mess.

**Comment by junpenglao** [comment]
I agree with @lucianopaz and this sounds like a good idea to me in general. However, we might want to separate the type hint part with the actual PR? Or is it the plan now to do this for each PR?

You should also do the same change for `sample_posterior_predictive_w`
https://github.com/pymc-devs/pymc3/blob/master/pymc3/sampling.py#L1114

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: I have found a way to import this symbol only during type checking, which seems to make pylint and mypy *both* happy.
Rebased and force-pushed this change for review.

**Comment by rpgoldman** [comment]
> I agree with @lucianopaz and this sounds like a good idea to me in general. However, we might want to separate the type hint part with the actual PR? Or is it the plan now to do this for each PR?

I see your point about the type hint part, but separating out the type hints will be quite burdensome (it's not something `diff` is good at).  Perhaps someone who has a better diff tool could split out the type hints as part of squash-merging this PR?

I use type hints both to provide valuable checking, and as a tool in figuring out PyMC3, which is a very complex code base; they have become an integral part of my Python coding process.

> You should also do the same change for `sample_posterior_predictive_w`

But this function doesn't have a `vars` argument at all, so I'm not sure what you would like to have done to it.  I'm happy to make whatever changes, but I need some more direction.

**Comment by junpenglao** [comment]
oh right sample_posterior_predictive_w only works for observed variable - never mind :-)

**Comment by rpgoldman** [comment]
Integrated the history, with blind alleys, into a single commit.
Retained the history with tag.

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py]: Why is this commented out?

File: pymc3/sampling.py
Comment (lucianopaz) [inline pymc3/sampling.py:2]: Nice

**Review by lucianopaz — COMMENTED** [review]
You still need to add this change to the bottom of the deprecations section of the release notes

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py:2]: 😀

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: It's commented out because the variable is set but never referenced, and pylint complained. Arguably it should not be in this commit, but it seems too trivial to deserve one of its own.  I could put it back if you like.

I wondered if this was actually trying to set a global that I didn't know about, but I didn't find any explanation -- maybe it's obsolete code that should go, or I suppose I might have missed something important.  If I'm wrong, and it *is* important, we should probably try to make a test that establishes its importance.

Comment (lucianopaz) [reply pymc3/sampling.py]: Ah, ok! I think this might have been left dangling after we removed live plots. Could you delete it?

**Comment by lucianopaz** [comment]
Once #3470 gets sorted out and merged, you'll have to rebase this PR to get the minimum syntax check of python3.5 working

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: Done and pushed.

**Comment by rpgoldman** [comment]
> Once #3470 gets sorted out and merged, you'll have to rebase this PR to get the minimum syntax check of python3.5 working

I subscribed to that PR. When I hear that it's merged, I will rebase and someone can merge this and close it out.

**Comment by rpgoldman** [comment]
> You still need to add this change to the bottom of the deprecations section of the release notes

Just did that and pushed.

**Review by lucianopaz — APPROVED** [review]
The PR looks fine now and could be merged. Do most of us agree that it's better to use `varnames` vs `vars`? I am in favor of `varnames` but I would like know the rest of us approved this change. @twiecki, @fonnesbeck, @junpenglao, @ColCarroll, and anyone else care to give their final opinion?

**Comment by twiecki** [comment]
I approve, it's more consistent. But didn't we switched to var_names?

**Comment by lucianopaz** [comment]
For now it's still mostly inconsistent but we made the switch to `var_names` in the plots. We can do it here too. Also, I just checked `sample_prior_predictive` and that uses `vars` too. It should be dealt with just like in `sample_posterior_predictive`

**Comment by twiecki** [comment]
Yeah let's move everything to var_names.

**Comment by rpgoldman** [comment]
> For now it's still mostly inconsistent but we made the switch to `var_names` in the plots. We can do it here too. Also, I just checked `sample_prior_predictive` and that uses `vars` too. It should be dealt with just like in `sample_posterior_predictive`

I have pushed the changes to sample_prior_predictive().  I am waiting to make sure they pass the tests. If so, I will do the `varnames` -> `var_names` change, and then I hope we can do the merge.

**Comment by rpgoldman** [comment]
OK, I think that this is all ready to go now.  `varnames` have become `var_names`.  If the tests all pass, I will make sure it's properly rebased and squash it down to a single commit for merging, unless there are any further issues?

File: pymc3/model.py
Comment (ColCarroll) [inline pymc3/model.py]: only `Optional` used here?

File: pymc3/sampling.py
Comment (ColCarroll) [inline pymc3/sampling.py]: `ValueError` for a bad argument here (`Exception` is too broad, I think)

File: pymc3/sampling.py
Comment (ColCarroll) [inline pymc3/sampling.py]: `ValueError` here, too

File: pymc3/sampling.py
Comment (ColCarroll) [inline pymc3/sampling.py]: does this one break on python 3.5? I thought we could only do signature type checks...

**Review by ColCarroll — COMMENTED** [review]
looks generally good! can you tighten up the exceptions before merge,though?

**Comment by rpgoldman** [comment]
I have a question about `sample_posterior_predictive` based on some tests I am doing.  According to the docstring, the `vars` argument need only be an `iterable`.  But is that sufficient, or if we pass a generator, is there a chance it will be emptied?

In line 1113 we draw values from `vars`, and in the following line we zip it together with a set of values.  Can't this fail if `vars` is a generator?  Does this need to be a list instead? Thanks!

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: Oh, probably. Sorry. I'll fix that.

Comment (rpgoldman) [reply pymc3/sampling.py]: Done.  See 99c7362d81bc729e0fa62de7039bb19b2a8d991d

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: See 5e546b252ec20b5b202037de99f377da91b3356c

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: See 5e546b252ec20b5b202037de99f377da91b3356c

File: pymc3/model.py
Comment (rpgoldman) [reply pymc3/model.py]: See 087ea2d8ebfaa2e34474eba0ce79f3420b78a4b4

**Comment by rpgoldman** [comment]
@ColCarroll @lucianopaz @twiecki I think I've fixed everything requested.  If the tests go fine, and you like the current version, I will squash this down (or you can do a squash merge), and I hope it will be done.

File: pymc3/sampling.py
Comment (ColCarroll) [inline pymc3/sampling.py]: raw `assert`s are considered bad form (they can be ignored if you use some flag that no one actually uses) -- `if data is None: raise AssertionError`

**Review by ColCarroll — COMMENTED** [review]
looks good to me once test pass! thanks for sticking with this @rpgoldman!

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: Fixed in 65439179b1abe7d8e14b2dbf01d257034f669e7e

Comment (rpgoldman) [reply pymc3/sampling.py]: @ColCarroll --
OK, I have no idea why this happened: when I replaced that line per above I now get a bunch of test failures involving failure to find `scipy.misc.factorial`  that look like this:

```
pymc3/tests/test_distributions.py:500: in check_dlogp
    assert_almost_equal(dlogp(pt), ndlogp(pt), decimal=decimals, err_msg=str(pt))
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:723: in __call__
    *args, **kwds).squeeze()
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:666: in __call__
    return super(Jacobian, self).__call__(np.atleast_1d(x), *args, **kwds)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:297: in __call__
    results = self._derivative(xi, args, kwds)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:663: in _derivative_nonzero_order
    return self._apply_fd_rule(step_ratio, results, steps2)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:394: in _apply_fd_rule
    fd_rule = self._get_finite_difference_rule(step_ratio)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:378: in _get_finite_difference_rule
    fd_mat = self._fd_matrix(step_ratio, parity, num_terms)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
step_ratio = 2.0, parity = 1, nterms = 1
    @staticmethod
    def _fd_matrix(step_ratio, parity, nterms):
        """
        Return matrix for finite difference and complex step derivation.
    
        Parameters
        ----------
        step_ratio : real scalar
            ratio between steps in unequally spaced difference rule.
        parity : scalar, integer
            0 (one sided, all terms included but zeroth order)
            1 (only odd terms included)
            2 (only even terms included)
            3 (only every 4'th order terms included starting from order 2)
            4 (only every 4'th order terms included starting from order 4)
            5 (only every 4'th order terms included starting from order 1)
            6 (only every 4'th order terms included starting from order 3)
        nterms : scalar, integer
            number of terms
        """
        _assert(0 <= parity <= 6,
                'Parity must be 0, 1, 2, 3, 4, 5 or 6! ({0:d})'.format(parity))
        step = [1, 2, 2, 4, 4, 4, 4][parity]
        inv_sr = 1.0 / step_ratio
        offset = [1, 1, 2, 2, 4, 1, 3][parity]
        c0 = [1.0, 1.0, 1.0, 2.0, 24.0, 1.0, 6.0][parity]
        c = c0 / \
>           misc.factorial(np.arange(offset, step * nterms + offset, step))
E       AttributeError: module 'scipy.misc' has no attribute 'factorial'
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:330: AttributeError
```

I do **not** see these on my laptop -- all the tests pass there.  This does not seem to have anything to do with my change to the `assert`, but what do I know?

Comment (ColCarroll) [reply pymc3/sampling.py]: my guess is a version of something (`numdifftools`?) changed under this diff. Comparing versions from the previous travis build and this one should let us know (and figure out if there's a good change). I'll dig into it after work.

Comment (rpgoldman) [reply pymc3/sampling.py]: That looks right -- I don't think it's anything I did. Is there some way to force a re-run of Travis on `master`, so that we can see if this failure also happens there (and my little tweak is innocent)?

Comment (rpgoldman) [reply pymc3/sampling.py]: @ColCarroll `numdifftools` didn't change, **but**  `scipy` 1.3.0 was released three days ago (17 May).  This looks like an upstream problem with `numdifftools`.  

And yes,`factorial` was removed from `scipy.misc`: https://github.com/scipy/scipy/commit/1cde30518fdc46c2c765beaa3e132e322ea7a579#diff-0763dd201fcbf9b088f9eb502b4f3d42

AFAICT this is fixed in the version of `numdifftools` on GitHub (or at least it is *different* -- factorial comes from `scipy.special` instead of `scipy.misc` now) *but* `numdifftools` has not seen a release in two years.

When you've got the factorial problem fixed, please squash merge this PR, or LMK if you would prefer me to smash it down to one commit for you to merge. Thanks!

**Comment by twiecki** [comment]
@rpgoldman What's your email?

**Comment by rpgoldman** [comment]
Rebased onto the latest master.  If the tests pass now, I will squash this down to one commit so there's a clean history.

**Comment by twiecki** [comment]
We can also just squash merge here.

**Comment by rpgoldman** [comment]
Pushed a cleaned up version of the history. Waiting for the checks, then I hope we can merge and close.

**Comment by lucianopaz** [comment]
Thanks @rpgoldman!

**Pull request merged**
